// rest-hooks/lib/test/providers.js
import _jsx from "@babel/runtime/helpers/esm/jsx";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import React from 'react';
import { reducer, ExternalCacheProvider, CacheProvider, PromiseifyMiddleware } from '..'; // Extension of the DeepPartial type defined by Redux which handles unknown

var makeExternalCacheProvider;

try {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  var _require = require('redux'),
      createStore = _require.createStore,
      applyMiddleware = _require.applyMiddleware;

  makeExternalCacheProvider = function makeExternalCacheProvider(managers, initialState) {
    var store = createStore(reducer, initialState, applyMiddleware.apply(void 0, _toConsumableArray(managers.map(function (manager) {
      return manager.getMiddleware();
    })).concat([PromiseifyMiddleware])));
    return function ConfiguredExternalCacheProvider(_ref) {
      var children = _ref.children;
      return _jsx(ExternalCacheProvider, {
        store: store,
        selector: function selector(s) {
          return s;
        }
      }, void 0, children);
    };
  };
} catch (e) {
  makeExternalCacheProvider = function makeExternalCacheProvider(managers, initialState) {
    throw new Error('Using makeExternalCacheProvider() requires redux to be installed as a peerDependency to rest-hooks');
  };
}

var makeCacheProvider = function makeCacheProvider(managers, initialState) {
  return function ConfiguredCacheProvider(_ref2) {
    var children = _ref2.children;
    return _jsx(CacheProvider, {
      managers: managers,
      initialState: initialState
    }, void 0, children);
  };
};

export { makeExternalCacheProvider, makeCacheProvider };
