// rest-hooks/lib/react-integration/NetworkErrorBoundary.js
import _jsx from "@babel/runtime/helpers/esm/jsx";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import React from 'react';

function isNetworkError(error) {
  return Object.prototype.hasOwnProperty.call(error, 'status');
}

var NetworkErrorBoundary =
/*#__PURE__*/
function (_React$Component) {
  _inherits(NetworkErrorBoundary, _React$Component);

  function NetworkErrorBoundary() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, NetworkErrorBoundary);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(NetworkErrorBoundary)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {});

    return _this;
  }

  _createClass(NetworkErrorBoundary, [{
    key: "render",
    value: function render() {
      if (!this.state.error) {
        return this.props.children;
      }

      return _jsx(this.props.fallbackComponent, {
        error: this.state.error
      });
    }
  }], [{
    key: "getDerivedStateFromError",
    value: function getDerivedStateFromError(error) {
      if (isNetworkError(error)) {
        return {
          error: error
        };
      }
    }
  }]);

  return NetworkErrorBoundary;
}(React.Component);

_defineProperty(NetworkErrorBoundary, "defaultProps", {
  fallbackComponent: function fallbackComponent(_ref) {
    var error = _ref.error;
    return _jsx("div", {}, void 0, error.status, " ", error.response && error.response.statusText);
  }
});

export { NetworkErrorBoundary as default };
