// rest-hooks/lib/react-integration/provider/middleware.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { useReducer, useMemo, useRef, useEffect } from 'react';
import usePromisifiedDispatch from './usePromisifiedDispatch'; // TODO: release as own library?

/** Redux-middleware compatible integration for useReducer() */

export default function createEnhancedReducerHook() {
  for (var _len = arguments.length, middlewares = new Array(_len), _key = 0; _key < _len; _key++) {
    middlewares[_key] = arguments[_key];
  }

  var useEnhancedReducer = function useEnhancedReducer(reducer, startingState) {
    var stateRef = useRef(startingState);

    var _useReducer = useReducer(reducer, startingState),
        _useReducer2 = _slicedToArray(_useReducer, 2),
        state = _useReducer2[0],
        realDispatch = _useReducer2[1];

    useEffect(function () {
      stateRef.current = state;
    }, [state]);
    var dispatchWithPromise = usePromisifiedDispatch(realDispatch, state);
    var outerDispatch = useMemo(function () {
      var _dispatch = function dispatch() {
        throw new Error("Dispatching while constructing your middleware is not allowed. " + "Other middleware would not be applied to this dispatch.");
      }; // closure here around dispatch allows us to change it after middleware is constructed


      var middlewareAPI = {
        getState: function getState() {
          return stateRef.current;
        },
        dispatch: function dispatch(action) {
          return _dispatch(action);
        }
      };
      var chain = middlewares.map(function (middleware) {
        return middleware(middlewareAPI);
      });
      _dispatch = compose(chain)(dispatchWithPromise);
      return _dispatch;
    }, [dispatchWithPromise]);
    return [state, outerDispatch];
  };

  return useEnhancedReducer;
}

var compose = function compose(fns) {
  return function (initial) {
    return fns.reduceRight(function (v, f) {
      return f(v);
    }, initial);
  };
};
