// rest-hooks/lib/react-integration/hooks/useFetcher.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { useContext, useRef, useCallback } from 'react';
import { isDeleteShape } from "../../resource";
import { DispatchContext } from "../context";
var SHAPE_TYPE_TO_RESPONSE_TYPE = {
  read: 'rest-hooks/receive',
  mutate: 'rest-hooks/rpc',
  "delete": 'rest-hooks/purge'
};

/** Build an imperative dispatcher to issue network requests. */
export default function useFetcher(fetchShape) {
  var throttle = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var dispatch = useContext(DispatchContext); // we just want the current values when we dispatch, so
  // box the shape in a ref to make react-hooks/exhaustive-deps happy

  var shapeRef = useRef(fetchShape);
  shapeRef.current = fetchShape;
  var fetchDispatcher = useCallback(function (params, body, updateParams) {
    var _shapeRef$current = shapeRef.current,
        fetch = _shapeRef$current.fetch,
        schema = _shapeRef$current.schema,
        type = _shapeRef$current.type,
        getFetchKey = _shapeRef$current.getFetchKey,
        options = _shapeRef$current.options;
    var responseType = SHAPE_TYPE_TO_RESPONSE_TYPE[type];
    var key = getFetchKey(params);
    var identifier = isDeleteShape(shapeRef.current) ? shapeRef.current.schema.getId(params) : key;
    var resolve = 0;
    var reject = 0;
    var promise = new Promise(function (a, b) {
      resolve = a;
      reject = b;
    });
    var meta = {
      schema: schema,
      responseType: responseType,
      url: identifier,
      throttle: throttle,
      options: options,
      resolve: resolve,
      reject: reject
    };

    if (updateParams) {
      meta.updaters = updateParams.reduce(function (accumulator, _ref) {
        var _ref2 = _slicedToArray(_ref, 3),
            toShape = _ref2[0],
            toParams = _ref2[1],
            updateFn = _ref2[2];

        return _objectSpread(_defineProperty({}, toShape.getFetchKey(toParams), updateFn), accumulator);
      }, {});
    }

    dispatch({
      type: 'rest-hooks/fetch',
      payload: function payload() {
        return fetch(params, body);
      },
      meta: meta
    });
    return promise;
  }, [dispatch, throttle]); // any is due to the ternary that we don't want to deal with in our implementation

  return fetchDispatcher;
}
