// rest-hooks/lib/react-integration/hooks/useSubscription.js
import { useContext, useEffect, useRef } from 'react';
import { DispatchContext } from "../context";

/** Keeps a resource fresh by subscribing to updates. */
export default function useSubscription(fetchShape, params) {
  var dispatch = useContext(DispatchContext);
  /*
  we just want the current values when we dispatch, so
  box the shape in a ref to make react-hooks/exhaustive-deps happy
   "Although useEffect is deferred until after the browser has painted, it’s guaranteed to fire before any new renders.
  React will always flush a previous render’s effects before starting a new update." - https://reactjs.org/docs/hooks-reference.html#useeffect
  */

  var shapeRef = useRef(fetchShape);
  shapeRef.current = fetchShape;
  useEffect(function () {
    if (!params) return;
    var _shapeRef$current = shapeRef.current,
        _fetch = _shapeRef$current.fetch,
        schema = _shapeRef$current.schema,
        getFetchKey = _shapeRef$current.getFetchKey,
        options = _shapeRef$current.options;
    var url = getFetchKey(params);
    dispatch({
      type: 'rest-hooks/subscribe',
      meta: {
        schema: schema,
        fetch: function fetch() {
          return _fetch(params);
        },
        url: url,
        frequency: options && options.pollFrequency
      }
    });
    return function () {
      dispatch({
        type: 'rest-hooks/unsubscribe',
        meta: {
          url: url,
          frequency: options && options.pollFrequency
        }
      });
    }; // serialize params
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch, params && fetchShape.getFetchKey(params)]);
}
