// rest-hooks/lib/state/SubscriptionManager.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _memoize from "lodash/memoize";

/** Handles subscription actions -> fetch or receive actions
 *
 * Constructor takes a SubscriptionConstructable class to control how
 * subscriptions are handled. (e.g., polling, websockets)
 */
var SubscriptionManager =
/*#__PURE__*/
function () {
  function SubscriptionManager(Subscription) {
    _classCallCheck(this, SubscriptionManager);

    _defineProperty(this, "subscriptions", {});

    _defineProperty(this, "Subscription", void 0);

    this.Subscription = Subscription;
    this.getMiddleware = _memoize(this.getMiddleware);
  }
  /** Ensures all subscriptions are cleaned up. */


  _createClass(SubscriptionManager, [{
    key: "cleanup",
    value: function cleanup() {
      for (var _url in this.subscriptions) {
        this.subscriptions[_url].cleanup();
      }
    }
    /** Called when middleware intercepts 'rest-hooks/subscribe' action.
     *
     */

  }, {
    key: "handleSubscribe",
    value: function handleSubscribe(action, dispatch) {
      var url = action.meta.url;

      if (url in this.subscriptions) {
        this.subscriptions[url].add(action.meta.frequency);
      } else {
        this.subscriptions[url] = new this.Subscription({
          schema: action.meta.schema,
          fetch: action.meta.fetch,
          frequency: action.meta.frequency,
          url: url
        }, dispatch);
      }
    }
    /** Called when middleware intercepts 'rest-hooks/unsubscribe' action.
     *
     */

  }, {
    key: "handleUnsubscribe",
    value: function handleUnsubscribe(action, dispatch) {
      var url = action.meta.url;

      if (url in this.subscriptions) {
        var empty = this.subscriptions[url].remove(action.meta.frequency);

        if (empty) {
          delete this.subscriptions[url];
        }
      } else if (process.env.NODE_ENV !== 'production') {
        console.error("Mismatched unsubscribe: ".concat(url, " is not subscribed"));
      }
    }
    /** Attaches Manager to store
     *
     * Intercepts 'rest-hooks/subscribe'/'rest-hooks/unsubscribe' to register resources that
     * need to be kept up to date.
     *
     * Will possibly dispatch 'rest-hooks/fetch' or 'rest-hooks/receive' to keep resources fresh
     *
     */

  }, {
    key: "getMiddleware",
    value: function getMiddleware() {
      var _this = this;

      return function (_ref) {
        var dispatch = _ref.dispatch;
        return function (next) {
          return function (action) {
            switch (action.type) {
              case 'rest-hooks/subscribe':
                _this.handleSubscribe(action, dispatch);

                return Promise.resolve();

              case 'rest-hooks/unsubscribe':
                _this.handleUnsubscribe(action, dispatch);

                return Promise.resolve();

              default:
                return next(action);
            }
          };
        };
      };
    }
  }]);

  return SubscriptionManager;
}();

export { SubscriptionManager as default };
