// rest-hooks/lib/index.js
import { Resource, SimpleResource, SuperagentResource, FetchShape, DeleteShape, ReadShape, MutateShape, Schema, SchemaList, SchemaDetail, SchemaOf, schemas } from './resource';
import NetworkManager from './state/NetworkManager';
import RIC from './state/RIC';
import PollingSubscription from './state/PollingSubscription';
import SubscriptionManager from './state/SubscriptionManager';
import reducer, { initialState } from './state/reducer';
import { useDenormalized } from './state/selectors';
import { useCache, useCacheLegacy, useFetcher, useRetrieve, useResource, useResourceLegacy, useSubscription, useMeta, useError, CacheProvider, useInvalidator, useResetter, ExternalCacheProvider, PromiseifyMiddleware, NetworkErrorBoundary, NetworkError } from './react-integration';
import useSelectionUnstable from './react-integration/hooks/useSelection';
import { AbstractInstanceType, FetchOptions, Method, State, FetchAction, ReceiveAction, RPCAction, PurgeAction, Dispatch, MiddlewareAPI, Middleware, Manager } from './types';
import { StateContext, DispatchContext } from './react-integration/context';
var __INTERNAL__ = {
  initialState: initialState,
  StateContext: StateContext,
  DispatchContext: DispatchContext,
  RIC: RIC
};
export { Resource, SimpleResource, SuperagentResource, CacheProvider, ExternalCacheProvider, PromiseifyMiddleware, useCache, useCacheLegacy, useFetcher, useRetrieve, useInvalidator, useResetter, useResource, useResourceLegacy, useSubscription, useMeta, useError, useSelectionUnstable, useDenormalized, NetworkManager, SubscriptionManager, PollingSubscription, reducer, NetworkErrorBoundary, schemas, __INTERNAL__ };
